Add NIST keywrap as a cipher mode

jack-fortanix opened the pull request:
## Description
Adds the NIST keywrap algorithms as AEAD cipher modes. Closes #2003 see also #1658

I had to make two fixes to the cipher tests; they were not resetting the key in the encryption direction (which it seems does not matter for GCM or CCM, but does for NIST keywrap). And, in the test where data is re-encrypted, the output length is compared with the original ciphertext length rather than the original plaintext length. These are identical for GCM/CCM (due to tag being modeled separately) but not for keywrap due to padding.

My understanding is Fortanix already has a corporate CLA with ARM.

## Status
**READY** and waiting review

## Requires Backporting
New feature so probably no.

## Todos
- [x] Tests
- [ ] Documentation
- [x] Changelog updated

I will update docs and changelog after you have reviewed for acceptance.

jack-fortanix commented:
Jenkins build is failing but I cannot see the output. If you let me know what is wrong I will fix it.

hanno-arm commented:
@jack-fortanix Jenkins failing because of a known timing test issue on FreeBSD only. That's as good as a pass.

jack-fortanix commented:
@RonEld Thanks for review. I have added a Changelog entry, split PR into two commits one with just the fixes to test and the rest for the keywrap impl, and added more test inputs that I believe cover all cases. If there is some additional situation that these are still not hitting just let me know.

RonEld commented on tests/suites/test_suite_cipher.kw.data:
```diff
@@ -0,0 +1,135 @@
+KW AES-128 wrap rfc 3394
```
please rename the file from `kw` to `nist_kw` as the module name

RonEld commented on tests/suites/test_suite_cipher.kw.data:
```diff
@@ -0,0 +1,135 @@
+depends_on:MBEDTLS_AES_C:MBEDTLS_NIST_KW_C
+auth_crypt_tv:MBEDTLS_CIPHER_AES_128_KWP:"00000000000000000000000000000000"::"":"":"446C037F831092B147C372616357BF7D":"":"":"00010203":0
+
+KWP AES-128 5 bytes
```
All the other tests use the word `byte` in singular

jack-fortanix commented:
Hi @RonEld I added a Changelog entry in the bugfix commit, and also added a number of tests taken from the NIST CAVS files. I did not add all of them due to it seemed like a lot of data, instead just added every 3rd test from those files. I can easily tweak the script I used to include more or fewer tests as desired. Anyway it now covers {KW,KWP} with all key sizes. I also addressed your other comments with regards to renaming the test suite and fixing the comment.

RonEld commented:
@jack-fortanix Thank you for addressing my comments.
Indeed this is a lot of test data.
There is no need to have *all* the test vectors from the NIST CAVS files. The idea is to cover all cases. ( e.g. Different input lengths. Too large lengths. Too short lengths. Different padding lengths. Invalid input length . etc..).
I don't see any tests for the unwrapping functionality though.
Please reduce the test data for same use case ( one "COUNT" for every input length is enough) and:
1. Add negative tests
2. Add unwrapping tests.
Thanks in advance

jack-fortanix commented:
OK I'll prune things down a bit. Re unwrapping, AIUI this is already being tested (the cipher test code first unwraps/decrypts, then wraps/encrypts to confirm identical outputs). However you are right there are no tests for unwrap failing or invalid lengths. I'll address this shortly.

jack-fortanix commented:
@RonEld I reduced the test cases as you suggested and added some failing tests, also from CAVS.

RonEld commented on tests/suites/test_suite_cipher.nist_kw.data:
```diff
@@ -0,0 +1,271 @@
+depends_on:MBEDTLS_AES_C:MBEDTLS_NIST_KW_C
+auth_crypt_tv:MBEDTLS_CIPHER_AES_128_KW:"7575da3a93607cc2bfd8cec7aadfd9a6":"":"":"031f6bd7e61e643df68594816f64caa3f56fabea2548f5fb":"":"":"42136d3c384a3eeac95a066fd28fed3f":0
+
+KW AES-128 wrap CAVS 17.4 WTF PLEN 16
```
nice touch:)
Please change to "FAIL"

jack-fortanix replied:
Haha oops script failure, will fix.

RonEld commented:
@jack-fortanix I have backported the bug fix on your behalf to `mbedtls-2.7` and `mbedtls-2.17` in #2519 and #2520

jack-fortanix commented:
@yanesca Any chance this will be reviewed soon? We are disappointed this didn't make it into the 2.17 release.

yanesca commented:
@jack-fortanix Unfortunately I have a quite long review queue. I'll try to prioritise this PR and will ask someone else for a review if I can't. Thank you for your patience.

jack-fortanix commented:
Addressed @yanesca's comments and rebased against latest `development`.

jack-fortanix commented:
Travis failures are in DTLS nbio test, I believe unrelated to this PR.

RonEld commented on tests/CMakeLists.txt:
```diff
@@ -130,11 +131,6 @@ if(NOT USE_CRYPTO_SUBMODULE)
     add_test_suite(rsa)
     add_test_suite(xtea)
 endif()
-add_test_suite(debug)
```
Why were these tests removed? IS it a rebase issue?

jack-fortanix replied:
Rebase error, sorry about that. Fixed now.

yanesca commented:
As @RonEld mentioned, we still need backports for the bugfix part before we could merge.

@jack-fortanix Could you please raise PRs targeting the mbedtls-2.7 and mbedtls-2.16 branches with the bugfix commit (1dbc5a257fd0767ca741ec18b951f1d666d3a062)?

yanesca commented:
@jack-fortanix Please don't mind my previous comment, @RonEld had already created the backport branches. Sorry for the noise.
